File: PRACTICA1/RUBY/Deepspace/Dice.rb
#encoding:utf-8
require './GameCharacter'
module Deepspace


# @author Lucía Salamanca López, Celia Arias Martínez

class Dice


 def initialize
   @NHANGARSPROB=0.25
   @NSHIELDSPROB=0.25
   @NWEAPONSPROB=0.33
   @FIRSTSHOTPROB=0.5
   @generator=rand
 end


attr_reader :NHANGARSPROB, :NSHIELDSPROB, :NWEAPONSPROB, :FIRSTSHOTPROB, :generator


 def initWithNHangars
   @generator=rand
   if (@generator > @NHANGARSPROB )
     return 1
   else
     return 0
   end
 end

 def initWithNWeapons
   @generator=rand
   if (@generator > 2*@NWEAPONSPROB)
     return 3
   else
     if (@generator < @NWEAPONSPROB)
       return 1
     else
       return 2
     end
   end
 end

 def initWithNShields
   @generator=rand
   if (@generator > @NSHIELDSPROB )
     return 1
   else
     return 0
   end
 end

 def whoStarts(nPlayers)
   rand(nPlayers)
 end


 def firstShot
   @generator=rand
   if (@generator > @FIRSTSHOTPROB )
     return GameCharacter::ENEMYSTARSHIP
   else
     return GameCharacter::SPACESTATION
   end

 end

 def spaceStationMoves(speed)
   @generator=rand
   if (@generator > speed)
     return false
   else
     return true
   end
 end

 def to_s
   "Este dado tiene de probabilidad de Hangar :"+@NHANGARSPROB.to_s+", de Shield: " +@NSHIELDSPROB.to_s+", de Weapons: " +@NWEAPONSPROB.to_s+", y de FirstShot: "+ @FIRSTSHOTPROB.to_s+"."
 end



end

end

File: PRACTICA1/JAVA/deepspace/src/deepspace/Dice.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.Random;

/**
 *
 * @author Lucía Salamanca López
 *         Celia Arias Martínez
 */
class Dice {
    private final float NHANGARSPROB;
    private final float NSHIELDSPROB;
    private final float NWEAPONSPROB;
    private final float FIRSTSHOTPROB;
    private final float EXTRAEFFICIENCYPROB;
    private Random generator;
    
    Dice(){
        generator = new Random(System.currentTimeMillis());
        NHANGARSPROB=0.25f;
        NSHIELDSPROB=0.25f;
        NWEAPONSPROB=0.33f;
        FIRSTSHOTPROB=0.5f;
        EXTRAEFFICIENCYPROB=0.8f;
    }
    
    int initWithNHangars(){
        if (generator.nextFloat() < NHANGARSPROB)
            return 0;
        return 1;
    }
    
    int initWithNWeapons(){
        float aleat = generator.nextFloat();
        if ( aleat < NWEAPONSPROB)
            return 1;
        else if(aleat < 2*NWEAPONSPROB)
            return 2;
        else 
            return 3;
    }
    
    int initWithNShields(){
        if (generator.nextFloat() < NSHIELDSPROB)
           return 0;
        return 1;
    }
    
    int whoStarts(int nPlayers){
        return generator.nextInt(nPlayers);
    }
    
    GameCharacter firstShot(){
        GameCharacter personaje;
        if (generator.nextFloat() < FIRSTSHOTPROB)
            personaje = GameCharacter.SPACESTATION; 
        else
            personaje = GameCharacter.ENEMYSTARSHIP;
        return personaje;    
    }
    
    boolean spaceStationMoves(float speed){
        if (generator.nextFloat() < speed)
           return true;
        return false;
    }
    
    public boolean extraEfficiency(){
        if (generator.nextFloat() < EXTRAEFFICIENCYPROB)
            return true;
        
        return false;
                    
    }
    
    public String toString(){
        String mensaje = "Este dado tiene de probabilidad de Hangar :"+NHANGARSPROB+", de Shield: " +NSHIELDSPROB+", de Weapons: " +NWEAPONSPROB+", y de FirstShot: "+ FIRSTSHOTPROB+".";
 
        return mensaje;
    }
}
